### PR TITLE
New picking method

### PR DESCRIPTION
**Pull request by Loloppe**
Use movement speed and item distance to pick item more quickly, no more click spam.
I didn't find how to get player Movement Speed directly, so it's configurable for now.

### CODE REVIEW

**Comment by Queuete** [comment]
I will try this out before I merge it, will most likely happen after the weekend, sorry.
Thanks for your contribution!

**Comment by Loloppe** [comment]
I've done some pretty big modification again, please make sure to try out the latest version. Thank you!

**Comment by Queuete** [comment]
Hello,

I played yesterday quite a few hours with it. 

From a player point of view:
Your updates make the pick up of lootexplosion more smooth (Harbingers, juiced maps where you loot after the map is done etc.), but looting mid-fighting / mid-mapping is more clunky. 

From a code point of view:
The problem to solve is the character wiggling around due to clicks which probably not hit the item label or simply to fast clicks. 

The "Simple" plugin philosophy includes 
- low amount of configuration
- low amount of data (offset) useage (to be available after a leaguestart as soon as possible) 
- short and easy to grasp code

I am not a fan of using movement speed. Having this in the configuration is bad, especially due to its volatile nature (quicksilver flask, onslaught, smoke mine other temporary effects). Reading this from the game is bad due to needing another offset.

Using the "Z" button to realign the items seems to be a good idea and is also a bearable addition to the configuration. In most cases the default value will be enough, whoever changed it will need to handle that himself. Furthermore it wont break the whole plugin if this is not configured correctly.

I dont see the reason for pulling the loop into the logic instead of handling 1 item at the time.
The only advantage is performance if the calculation of the closest label would take too long. I doubt this difference is noticeable at all.
It adds complexity to the logic and the problem that the items to pick are from the point in time when the button was first pressed instead of up to date.

In summary: I wont merge this in the current state. I think the problem you are solving is an important one, but the current implementation has quite a negative impact on another totally valid type of use and the implementation is rather complex. 

Maybe instead of using movement speed we could track if the next item to pick is the same as the last one and increase the delay of clicks in those cases? I probably give something like that a try within the next days.

**Comment by Loloppe** [comment]
I understand. The main reason why I've made it take a List of item and not doing it one item at a time is because it was extremely slow in juiced map mostly. I will try to compare it with another implementation that doesn't do that and simplify the code if possible.

For example, picking each item was taking almost one second on the old version. Meanwhile in this version it can pick 3-4 item per second.

If it's still too slow, the plugin could have two method that can be switched between in the settings. (One for normal gameplay and the current one for heavier situation)

As for the movement speed integration, it's currently used only to break out of the loop if something wrong happen (the item doesn't get picked). That "failsafe" could be fixed to a specific duration, which would be less calculation, but also slower if there's mistake being done, etc. I could add it to the settings as something configurable.

**Comment by Loloppe** [comment]
So I've done some some tests and having to loop through all labels so many time is just way too slow to my taste. I agreed with your point that "It adds complexity to the logic and the problem that the items to pick are from the point in time when the button was first pressed instead of up to date." but at the same time, the looting speed should be the main priority of this plugin. 

For example, even if a new item drop while your looting, after the plugin is done looting the current List of items or if it can't reach the next item in it List, it will detect and loot the new items right after and still be way faster. Currently, the more labels there is, the more the plugin struggle and slow down in speed. 

I've simplified as much as I could the code. I've removed the movement speed and set a static value that can be configured.

As for creating a List of items into a loop to grab them, I'm not sure if it's because we're checking all labels at all time, or maybe it's because that we need to wait for the Tick of the HUD every time we want to pick an item?, but it's just so slow in my case, especially in labels-heavy situation, but also in normal situation.

I'm not sure what else I could do about it, but it definitely feel better to use now.
It still use the same offset as before.
There's a bit more thing that can be configured, but it's very simple stuff and it's easy to grasp.
As for the code, it could be simplified even further, but it wouldn't really affect the performance.

**Comment by Loloppe** [comment]
I think my fork is mostly stable now, I haven't had that NullReferenceException after a few maps.
One thing that I found out is that it often skip one item (probably from it being removed from the list while not being picked), I'm not really sure on how to fix that without considerably slowing the speed down, so might as well just keep it like it is right now. The item end up being picked either way after the list reset.

**Comment by Loloppe** [comment]
I rarely and randomly get ThreadAbortException from the wait loop, I'm not sure if it's because of how ExileApi handle it Job but I got no idea how to fix that. Your plugin SimpleStashies also does that some time. I'm also unsure if there's a way to just restart a Job if it crash? There's like no button on ExileApi to reload plugin or something.

**Comment by Queuete** [comment]
Hey @Loloppe,
sorry for not answering earlier - it seems I had notifications only enabled for the ExileApi repository but not this one.
I tested your changes shortly before looking into the Coroutine changes and multiple times the plugin just stopped working (probably the ThreadAbortException you are talking about?).
I am not 100% sure, but the Thread most likely gets aborted when its running too long. This is mostly the case when you use the "Thread.Sleep(ms)" to pause the execution. This should not happen when using the Coroutines and "yield return new Waittime(ms)"
My current implementation definitly has some flaws and there will be a few more changes in the evening and some of the changes definitly are inspired by your fork and I am thankful for your contribution even though it did not end up in the current state.
